### ISSUE TITLE
Feature Request: Provide global Replace option to consolidated Mesh & Texture file paths into a different folder

### ISSUE THREAD

**Issue by RuneLore63** [TES5Edit/TES5Edit#1047]
Currently there is no option / script for global relocation / replacement of file paths while reorganising mod structures. This generally affects meshes and textures. Yet, some mods include naked armour sets pointing to body part & head "nif" files with associated eye, mouth, & hair texture "dds" files. Hence, reworking a series of mods (e.g. FollowerSet1, FollowerSet2 & FollowerSet3) into a consolidated folder requires numerous arduous manual edits.

If for example: FollowerSet1 has Arthur, Benjamin & Charles; FollowerSet2 has David, Edward & Frederick; while FollowerSet3 has Gerald, Horace & Ian; then the actors\character file path has 1 subfolder for each character (9 subfolders in total) whereas a simple replacement of "actors\character\charactername\ " with "actors\character\NEWFOLDER\charactername\ " containing "dds" files and "actors\character\charactername\ " with "actors\character\NEWFOLDER\charactername\ " containing "nif" files would speed up the process.

Generally global file path replacement only needs to be done for Armor Addons, Weapons, Head Parts, Body Parts & Texture Sets. But, it would also be helpful if there was an option to allow replacement of fixed bodies to using the games standard body nif files and/or texture dds files. As well as, if the option could move the old path folders into the NewFolder as subfolders.

While the 'NIF/BGSM/BGEM Batch Textures Replacement' script appears to partially do what is needed, that's for MorrowWind. However, processing of the unpacking of archive files into loose files would also be necessary. Is this possible?

**Comment by jbostrus** [comment]
TL;DR Text search and replace at a core level isn't the most easily implemented thing. A script that is more aware of specifics to what you're trying to search and replace is really the only viable choice.

Depending on the game there are multiple factors that come into play here. What you propose is only viable in concept if the changes were made before ever publishing a mod, and done by the original mod author. A user of a mod should never be doing these kinds of changes. What you describe is essentially just a search and replace of text. Which is entirely possible via scripts as you started to mention. Scripts can also copy assets from a loaded archive out to loose files. But this should be a non issue if you are the original author as you should already have the loose files for the production.

Some of what comes may be redundant for you but I put it here for others who may not be familiar to better understand what's going on. I'll speak directly about Skyrim and Fallout 4 as I'm most familiar with those. But most, if not all, previous games to Skyrim work similar to it.

For either of them the path to nif files is arbitrary. It doesn't have to follow any naming scheme. And the paths don't matter what so ever beyond ease of locating by the author of the plugin. You could drop all your nif files in the root of data/meshes and call it a day. This would a horrible design choice, if for no other reason than you might have conflicts with other mods if those authors chose to do the same. The issue comes as name collisions if you and the other mod use the same file names. But it's technically valid.

I suspect Bethesda chose to use a hierarchy scheme for asset management so their designers had some general idea of where things **should** be put in a hierarchy to make it easier for other designers to find them. Though even at that they didn't hold strictly consistent to their own choices.

At it's core xEdit was purposed for viewing, and the later editing of plugin files. So changing the paths of the text fields pointing to a nif is easy and basic. Since there's no hard definition of which subrecord signatures are text and which aren't a search and replace at the global level as you put it isn't really easy. Scripts have already been made to do that sort of thing.

To compound the technical challenges, texture paths are set and embedded within the nif, and in the case of Fallout 4 also material files (bgsm/bgem). Fallout 4 nifs can contain the texture pathing information in a shader node directly in the nif and/or via a referenced material file. So moving things around requires editing multiple different files: plugin, nif,(s) and material(s). Which is not the original design of the tool.

You can then compound the issue further in the case that there are other mods, patchers or replaceers, which might reference the old path after you've changed it. So now you would have more conflicting stuff. Which brings me back to my earlier statement: **What you propose is only viable in concept if the changes were made before ever publishing a mod, and done by the original mod author.**

Now with all that said the scripting system has additional code support to do fancy things that the core wasn't designed to do. Such as stuff that isn't directly plugin related. There are some included scripts with xEdit that will do what you want but you need to edit the search and replace data to fit your particular needs. You mentioned one of them. I don't know why you think it only worked for MorrowWind as I didn't see anything in it that specifically stated such.

```
NIF - Batch textures replacement.pas
Replace model file name.pas
Replace model alternate textures.pas

```
The one thing conspicuously absent is something to update the texture set paths. The above mentioned script for alternate textures changes which TXST is pointed to, but not the data within the TXST that references the actual dds file.

As another search and replace option you could also try StringOps by fireundubh.
https://github.com/fireundubh/xedit-scripts/blob/master/gui/StringOps.pas
I believe if you enter `[*]` as the path it will check all fields within a record instead of just the one specified by the path. I've not personally used it myself so that might be wrong. But keep in mind StringOps specifically only handles the plugin changes. Not any of the external asset file changes that would be necessary.

Ultimately for your very specific use case scenario a custom script written as a combination of the previously mentioned ones is probably the best choice.

**Comment by RuneLore63** [comment]
Currently brushing up on my Delphi skills (haven't coded in that for 20 years) & as I expected, it really requires a script!

Agreed ... I saw mention of MorrowWind in the script code and thought that script was for that, but on re-reading it, it's (including that). Cataracts & old age don't help with such things. However, while I acknowledge everything you've said, it still amazing that records already loaded into xEdit cannot have basic find / replace functionality run on them and require manual editing of every entry.

Agreed ... previously  published mods with patches would be more of a nightmare to work with. Hence, why I specifically mentioned follower mods as an example, which do unfortunately do have clashing character names at times - which, currently require manual movement to an "actors\character\CharacterName\ " folder . Personally, I am more interested in Skyrim SE initially (as that's what I produce / amend mods for) and then maybe look at FO4, etc.

Thanks very much for the StringOps.pas reference, I'm looking at that currently.
